devito/finite_differences/differentiable.py

```diff
@@ -186,20 +187,93 @@ def _has(self, pattern):
         return super(Differentiable, self)._has(pattern)
 
 
-class Add(sympy.Add, Differentiable):
-    pass
+class DifferentiableOp(Differentiable):
 
+    def __new__(cls, *args, **kwargs):
+        obj = cls.__base__.__new__(cls, *args, **kwargs)
 
-class Mul(sympy.Mul, Differentiable):
-    pass
+        # Unfortunately SymPy may build new sympy.core objects (e.g., sympy.Add),
+        # so here we have to rebuild them as devito.core objects
+        obj = diffify(obj)
 
+        return obj
 
-class Pow(sympy.Pow, Differentiable):
-    pass
 
+class Add(sympy.Add, DifferentiableOp):
+    __new__ = DifferentiableOp.__new__
 
-class Mod(sympy.Mod, Differentiable):
-    pass
+
+class Mul(sympy.Mul, DifferentiableOp):
+    __new__ = DifferentiableOp.__new__
+
+
+class Pow(sympy.Pow, DifferentiableOp):
+    __new__ = DifferentiableOp.__new__
+
+
+class Mod(sympy.Mod, DifferentiableOp):
+    __new__ = DifferentiableOp.__new__
+
+
+class diffify(object):
+
+    """
+    Helper class based on single dispatch to reconstruct all nodes in a sympy
+    tree such they are all of type Differentiable.
+
+    Notes
+    -----
+    The name "diffify" stems from SymPy's "simpify", which has an analogous task --
+    converting all arguments into SymPy core objects.
+    """
+
+    def __new__(cls, obj):
+        args = [diffify._doit(i) for i in obj.args]
+        obj = diffify._doit(obj, args)
+        return obj
+
+    def _doit(obj, args=None):
+        cls = diffify._cls(obj)
+        if cls is obj.__class__:
+            return obj
+
+        args = args or obj.args
+        newobj = cls(*args, evaluate=False)
+
+        # SymPy objects pretend to be immutable, but in reality they are not.
+        # As facts are deduced, a SymPy core object's ``_assumptions`` data
+        # structure is updated. Since here we use ``evaluate=False`` all over
+        # the place, we have to explicitly reassign the ``_assumptions``.
+        newobj._assumptions = obj._assumptions.copy()
```

Comment:
I think this would also go with `_from_args' but need to double check

Reply:
if it goes from `_from_args` it would be awesome, I'll check that

Reply:
https://github.com/sympy/sympy/blob/master/sympy/core/operations.py#L70

only is_commutative ? 😮

Reply:
Oh dang inded, I checked `rawargs` and same only the commutative is saved, that's not great